### ISSUE TITLE
`save_intermediary_cubes: true` option in config-user makes the whole show crash and burn when preprocessors that use masked cell_measures

### ISSUE THREAD

**Issue by valeriupredoi** [ESMValGroup/ESMValCore#1917]
### Reproduce with

Reproducible example:

```yaml
documentation:
  title: CMUG diagnostics for sea surface salinity

  description: |
    Recipe for evaluating models against ESACCI Sea Surface Salinity dataset

  authors:
    - vegas-regidor_javier

  maintainer:
    - vegas-regidor_javier

  references:
    - contact_authors

  projects:
    - cmug

preprocessors:
  timeseries:
    area_statistics:
      operator: mean
      fx_variables:
        areacello:

datasets:
  - {project: CMIP6, exp: historical, dataset: MPI-ESM1-2-HR, ensemble: r1i1p1f1,
      start_year: 1950, end_year: 2014, alias: MPI-ESM1-2-HR}

diagnostics:
  compare_salinity:
    description: ESACCI-SEA-SURFACE-SALINITY check
    variables:
      sos:
        preprocessor: timeseries
        mip: Omon
        grid: gn
    scripts: null
```
with `save_intermediary_cubes: true` in config-user.

### Error to reproduce

```
2023-02-02 16:01:57,860 UTC [10802] ERROR   Program terminated abnormally, see stack trace below for more information:
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/valeriu/ESMValCore/esmvalcore/_task.py", line 795, in _run_task
    output_files = task.run()
  File "/home/valeriu/ESMValCore/esmvalcore/_task.py", line 262, in run
    self.output_files = self._run(input_files)
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/__init__.py", line 643, in _run
    product.apply(step, self.debug)
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/__init__.py", line 436, in apply
    save(self.cubes, filename)
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/_io.py", line 337, in save
    iris.save(cubes, **kwargs)
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/site-packages/iris/io/__init__.py", line 482, in save
    saver(source, target, **kwargs)
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/site-packages/iris/fileformats/netcdf/saver.py", line 2754, in save
    sman.write(
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/site-packages/iris/fileformats/netcdf/saver.py", line 759, in write
    self._add_cell_measures(cube, cf_var_cube, cube_dimensions)
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/site-packages/iris/fileformats/netcdf/saver.py", line 1109, in _add_cell_measures
    return self._add_inner_related_vars(
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/site-packages/iris/fileformats/netcdf/saver.py", line 1053, in _add_inner_related_vars
    cf_name = self._create_generic_cf_array_var(
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/site-packages/iris/fileformats/netcdf/saver.py", line 1938, in _create_generic_cf_array_var
    raise ValueError(msg)
ValueError: Cell measures with missing data are not supported.
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/valeriu/ESMValCore/esmvalcore/_main.py", line 486, in run
    fire.Fire(ESMValTool())
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/site-packages/fire/core.py", line 466, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/site-packages/fire/core.py", line 681, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/home/valeriu/ESMValCore/esmvalcore/_main.py", line 393, in run
    self._run(recipe, session)
  File "/home/valeriu/ESMValCore/esmvalcore/_main.py", line 433, in _run
    process_recipe(recipe_file=recipe, session=session)
  File "/home/valeriu/ESMValCore/esmvalcore/_main.py", line 127, in process_recipe
    recipe.run()
  File "/home/valeriu/ESMValCore/esmvalcore/_recipe/recipe.py", line 1900, in run
    self.tasks.run(max_parallel_tasks=self._cfg['max_parallel_tasks'])
  File "/home/valeriu/ESMValCore/esmvalcore/_task.py", line 724, in run
    self._run_parallel(max_parallel_tasks)
  File "/home/valeriu/ESMValCore/esmvalcore/_task.py", line 767, in _run_parallel
    _copy_results(task, running[task])
  File "/home/valeriu/ESMValCore/esmvalcore/_task.py", line 790, in _copy_results
    task.output_files, task.products = future.get()
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/multiprocessing/pool.py", line 774, in get
    raise self._value
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/valeriu/ESMValCore/esmvalcore/_task.py", line 795, in _run_task
    output_files = task.run()
  File "/home/valeriu/ESMValCore/esmvalcore/_task.py", line 262, in run
    self.output_files = self._run(input_files)
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/__init__.py", line 643, in _run
    product.apply(step, self.debug)
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/__init__.py", line 436, in apply
    save(self.cubes, filename)
  File "/home/valeriu/ESMValCore/esmvalcore/preprocessor/_io.py", line 337, in save
    iris.save(cubes, **kwargs)
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/site-packages/iris/io/__init__.py", line 482, in save
    saver(source, target, **kwargs)
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/site-packages/iris/fileformats/netcdf/saver.py", line 2754, in save
    sman.write(
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/site-packages/iris/fileformats/netcdf/saver.py", line 759, in write
    self._add_cell_measures(cube, cf_var_cube, cube_dimensions)
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/site-packages/iris/fileformats/netcdf/saver.py", line 1109, in _add_cell_measures
    return self._add_inner_related_vars(
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/site-packages/iris/fileformats/netcdf/saver.py", line 1053, in _add_inner_related_vars
    cf_name = self._create_generic_cf_array_var(
  File "/home/valeriu/miniconda3/envs/esmvaltool-latest/lib/python3.10/site-packages/iris/fileformats/netcdf/saver.py", line 1938, in _create_generic_cf_array_var
    raise ValueError(msg)
ValueError: Cell measures with missing data are not supported.
2023-02-02 16:01:57,870 UTC [10802] INFO    
If you have a question or need help, please start a new discussion on https://github.com/ESMValGroup/ESMValTool/discussions
If you suspect this is a bug, please open an issue on https://github.com/ESMValGroup/ESMValTool/issues
To make it easier to find out what the problem is, please consider attaching the files run/recipe_*.yml and run/main_log_debug.txt from the output directory.
```

### Reasons and stuff

- `areacello` and a bunch other CMIP6 fx or cell measures have masked data (about 30-60% of the data is masked)
- the reason why this is happening only when saving intermediary cubes is set to true is that the regular case (when not saving intermediary cubes) benefits from a last preprocessing set that removes the fx variable right before saving
- to be honest, I am not 100% sure why iris are throwing thix esception instead of nicely reading in the `fill_valu` which is set at save point, I will open an issue and ask them
- we either say one can't use the intermediary saver in these cases, or we fix the thing proper - I guess a removal of the fx variable before the intermediary save would do fin - open to opinions

**Comment by valeriupredoi** [comment]
fortunately enough @bouweandela 's https://github.com/ESMValGroup/ESMValCore/pull/1609 solves this problem!

**Comment by bouweandela** [comment]
I suspect that is because it is no longer attaching the ancillary variable or attaching a different one. This might actually be a bug. Could you check why it works now?

**Comment by valeriupredoi** [comment]
I'll have a looksee today, bud :+1:

**Comment by valeriupredoi** [comment]
shoot! You were right @bouweandela - it is a bug, a couple of them by the looks of it. So use this recipe (same as above, but for the sake of not scrolling up):
```yaml
documentation:
  title: CMUG diagnostics for sea surface salinity

  description: |
    Recipe for evaluating models against ESACCI Sea Surface Salinity dataset

  authors:
    - vegas-regidor_javier

  maintainer:
    - vegas-regidor_javier

  references:
    - contact_authors

  projects:
    - cmug

preprocessors:
  timeseries:
    area_statistics:
      operator: mean
      fx_variables:
        areacello:

datasets:
  - {project: CMIP6, exp: historical, dataset: MPI-ESM1-2-HR, ensemble: r1i1p1f1,
      start_year: 1950, end_year: 2014, alias: MPI-ESM1-2-HR}

diagnostics:
  compare_salinity:
    description: ESACCI-SEA-SURFACE-SALINITY check
    variables:
      sos:
        preprocessor: timeseries
        mip: Omon
        grid: gn
    scripts: null
```

with `save_intermediary_cubes: true`

The output in `esmvaltool_output/recipe_mpmoine_20230217_144847/preproc/compare_salinity/sos` looks like this:

```
(esmvaltool-latest) valeriu@valeriu-PORTEGE-Z30-C:~$ ls -la esmvaltool_output/recipe_mpmoine_20230217_144847/preproc/compare_salinity/sos/*
-rw-rw-r-- 1 valeriu valeriu 47011 Feb 17 14:53 esmvaltool_output/recipe_mpmoine_20230217_144847/preproc/compare_salinity/sos/CMIP6_MPI-ESM1-2-HR_Omon_historical_r1i1p1f1_sos_gn_1950-2014.nc
-rw-rw-r-- 1 valeriu valeriu 83025 Feb 17 14:53 esmvaltool_output/recipe_mpmoine_20230217_144847/preproc/compare_salinity/sos/CMIP6_MPI-ESM1-2-HR_Omon_historical_r1i1p1f1_sos_gn_1950-2014_provenance.xml
-rw-rw-r-- 1 valeriu valeriu   850 Feb 17 14:53 esmvaltool_output/recipe_mpmoine_20230217_144847/preproc/compare_salinity/sos/metadata.yml

esmvaltool_output/recipe_mpmoine_20230217_144847/preproc/compare_salinity/sos/CMIP6_MPI-ESM1-2-HR_Ofx_historical_r1i1p1f1_areacello_gn_fixed:
total 8
drwxrwxr-x 2 valeriu valeriu 4096 Feb 17 14:49 .
drwxrwxr-x 5 valeriu valeriu 4096 Feb 17 14:53 ..

esmvaltool_output/recipe_mpmoine_20230217_144847/preproc/compare_salinity/sos/CMIP6_MPI-ESM1-2-HR_Omon_historical_r1i1p1f1_sos_gn_1950-2014:
total 56
drwxrwxr-x 2 valeriu valeriu  4096 Feb 17 14:49 .
drwxrwxr-x 5 valeriu valeriu  4096 Feb 17 14:53 ..
-rw-rw-r-- 1 valeriu valeriu 47011 Feb 17 14:50 00_area_statistics.nc

esmvaltool_output/recipe_mpmoine_20230217_144847/preproc/compare_salinity/sos/CMIP6_MPI-ESM1-2-HR_Omon_historical_r1i1p1f1_sos_gn_1950-2014_fixed:
total 8
drwxrwxr-x 2 valeriu valeriu 4096 Feb 17 14:49 .
drwxrwxr-x 5 valeriu valeriu 4096 Feb 17 14:53 ..
```

so it stops writing intermediary files after 00_ but for some reason it considers 00 to be the last step, area_statistics

**Comment by valeriupredoi** [comment]
ie concatenate, fix_metadata, fix_data, add_supplementary_variables etc are not writing intermediary files anymore - the Ofx variable addition (areacello) is happening, that's for sure, but if the intermediary file is not saved anymore, then the iris error doesn't happen

**Comment by valeriupredoi** [comment]
@bouweandela your latest commits have now brought back the intermediary cubes saving - well done! Also, the issue in this issue is still not reproducible with `split-recipe`, even though the Ofx variable is added, and handled correctly in area statistics - this is happening bc the intermediary files have now changed, in this setup we save:
```
00_load.nc
01_fix_metadata.nc
02_concatenate.nc
03_cmor_check_metadata.nc
04_clip_timerange.nc
05_fix_data.nc
06_cmor_check_data.nc
07_area_statistics.nc
```
the Ofx variable is added somewhere before the cmor check, and that takes care of the missing values that iris had issues saving. Or, at least, that's what I think it's happening :+1:

**Comment by bouweandela** [comment]
Unfortunately, I forgot to save the intermediary cube after adding the ancillaries. It should work (i.e. same old problem) with [05b1e30](https://github.com/ESMValGroup/ESMValCore/pull/1609/commits/05b1e30282d86276f4959a40f6150bdbc3d5305d).

**Comment by valeriupredoi** [comment]
that shall be tested on Monday, bud, :beer: o'clock now. Have a good weekend!

**Comment by valeriupredoi** [comment]
OK @bouweandela tested again, and this time around the familiar error does indeed pop up - the table of saved int files is like it used to be too:
```
00_load.nc
01_fix_metadata.nc
02_concatenate.nc
03_cmor_check_metadata.nc
04_clip_timerange.nc
05_fix_data.nc
06_cmor_check_data.nc
07_add_supplementary_variables.nc
```

iris folk said they'd remove this roadblock very soon, so we should be good :+1:

**Comment by valeriupredoi** [comment]
TL;DR (lots of intermediary guff, pun intended) - @pp-mo mentioned the iris folk would be glad to remove the check and raise in https://github.com/SciTools/iris/issues/5147#issuecomment-1422720537

**Comment by valeriupredoi** [comment]
was just about to close this, but alas, this is still happening with our latest Core 2.8 and iris=3.4.1, I pinged the iris folk https://github.com/SciTools/iris/issues/5147#issuecomment-1547840443

**Comment by valeriupredoi** [comment]
> was just about to close this, but alas, this is still happening with our latest Core 2.8 and iris=3.4.1, I pinged the iris folk [SciTools/iris#5147 (comment)](https://github.com/SciTools/iris/issues/5147#issuecomment-1547840443)

jumped the version gun, with iris=3.5 (already installed centrally on JASMIN by yours truly) all works biutifully so this can indeed be closed :+1:
